### PR TITLE
Nullable: System.Xml, part 3

### PR DESCRIPTION
**Pull request by krwq**
Mostly schema but not exclusively

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs [hunk 1]
```diff
@@ -19,20 +20,20 @@ internal enum Use
...
-        protected string defaultValueRaw;       // default value in its original form
-        protected object defaultValueTyped;
+        protected string? defaultValueRaw;       // default value in its original form
+        protected object? defaultValueTyped;
 
         protected long maxLength; // dt:maxLength
         protected long minLength; // dt:minLength
 
-        protected List<string> values;    // array of values for enumerated and notation types
+        protected List<string> values = null!;    // array of values for enumerated and notation types
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs:36]: I'm assuming you're forcing this because all inheriting types are initializing the field? If that's the case, would it maybe make sense to slightly refactor so that the constructor accepts it as a parameter?

Or alternatively leave as nullable and assert non-nullability on the two Check methods that reference it?

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentExtension.cs [hunk 2]
```diff
@@ -2,16 +2,17 @@
...
     using System.Collections;
     using System.Xml.Serialization;
 
     public class XmlSchemaComplexContentExtension : XmlSchemaContent
     {
-        private XmlSchemaParticle _particle;
-        private XmlSchemaObjectCollection _attributes = new XmlSchemaObjectCollection();
-        private XmlSchemaAnyAttribute _anyAttribute;
+        private XmlSchemaParticle? _particle;
+        private XmlSchemaObjectCollection? _attributes = new XmlSchemaObjectCollection();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentExtension.cs:14]: Looks like this can be non-nullable. Same for its corresponding property getter.

Unless the `SetAttributes` should be updated to accept nullable arguments.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs [hunk 3]
```diff
@@ -2,16 +2,17 @@
...
     using System.Collections;
     using System.Xml.Serialization;
 
     public class XmlSchemaComplexContentRestriction : XmlSchemaContent
     {
-        private XmlSchemaParticle _particle;
-        private XmlSchemaObjectCollection _attributes = new XmlSchemaObjectCollection();
-        private XmlSchemaAnyAttribute _anyAttribute;
+        private XmlSchemaParticle? _particle;
+        private XmlSchemaObjectCollection? _attributes = new XmlSchemaObjectCollection();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs:14]: Can be non-nullable.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs [hunk 4]
```diff
@@ -78,23 +79,24 @@ public static string EncodeLocalName(string name)
 
         public static string DecodeName(string name)
         {
-            if (name == null || name.Length == 0)
+            if (string.IsNullOrEmpty(name))
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:82]: This implies that the argument and return type should be nullable (and potentially marked with the NotNullIfNotNull attribute).

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs [hunk 5]
```diff
@@ -135,18 +135,16 @@ public static string ToString(string name, string ns)
         }
 
         // --------- Some useful internal stuff -----------------
-        internal void Init(string name, string ns)
+        internal void Init(string? name, string? ns)
         {
-            Debug.Assert(name != null && ns != null);
-            _name = name;
-            _ns = ns;
+            _name = name ?? string.Empty;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:140]: Why the change of contract here?

File: src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs [hunk 6]
```diff
@@ -277,7 +280,7 @@ public override int GetCharCount(byte[] bytes)
 
         public override Encoder GetEncoder()
         {
-            return null;
+            return null!;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs:283]: Depending on how pervasive this particular subtype is, it might indicate that the base class might need to be annotated differently.

**Review by eiriktsarpalis — APPROVED** [review]
Other than a few minor issues, lgtm.

File: src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.Common.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.Common.cs:5]: to get it off the radar since I don't do them in order, I'll bulk remove this in the end when adding property to the whole project

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/BitSet.cs [hunk 8]
```diff
@@ -195,10 +198,7 @@ public override bool Equals(object obj)
 
         public BitSet Clone()
         {
-            BitSet newset = new BitSet();
-            newset._count = _count;
-            newset._bits = (uint[])_bits.Clone();
-            return newset;
+            return new BitSet(_count, (uint[])_bits.Clone());
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/BitSet.cs:201]: perhaps - readonly is useful when writing new code, this code is unlikely to change anytime soon

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs (hunk 1 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs:36]: Generally this is an abstract type which derived types should fill this field as non-null since all call-sites expect it. Nullable feature  doesn't have a way to express it other than suppressing the warning. Since this is internal type I don't think it's worth adding `!` or Debug.Assert all callsites especially given how large this code base is.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs:14]: This is a public API which have corresponding property which allow to set it to null, I think this is more correct representation of state - assuming all callsites expect it to be non-null should we annotate this based on what makes more sense or to represent the state? I.e. we've annotated ToString as nullable even though 99.9% of the code assumes it's not null

File: src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs (hunk 5 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlQualifiedName.cs:140]: I changed it since all constructors accepted null and converted that to non-null - I think one of the callsites was passing nullable to Init which would make the `_name` nullable which would pretty much always throw NRE on all methods so instead of that I decided to just copy behavior from the constructor

File: src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs (hunk 6 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs:283]: This case was weird - all callsites which used this method expected this to be not-null and this type is never used directly so we ended up never calling this specific implementation returning null - I've decided to suppress this since this type is not public.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs [hunk 9]
```diff
@@ -428,44 +442,48 @@ internal static string VerifyNCName(string name, ExceptionType exceptionType)
...
 
         /// <devdoc>
         ///    <para>
         ///    </para>
         /// </devdoc>
-        public static string VerifyTOKEN(string token)
+        public static string? VerifyTOKEN(string token)
         {
-            if (token == null || token.Length == 0)
+            if (string.IsNullOrEmpty(token))
```
Comment (buyaa-n) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:463]: NIT: accepts null and can have `NotNullIfNotNull`

File: src/libraries/System.Private.Xml/src/System/Xml/XmlSecureResolver.cs [hunk 10]
```diff
@@ -22,12 +23,12 @@ public override ICredentials Credentials
             set { _resolver.Credentials = value; }
```
Comment (buyaa-n) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlSecureResolver.cs:23]: Above on row 16, maybe make the not used `securityUrl` nullable?

File: src/libraries/System.Private.Xml/src/System/Xml/XmlSecureResolverAsync.cs [hunk 11]
```diff
@@ -2,13 +2,14 @@
...
 
+#nullable enable
 using System.Threading.Tasks;
 
 namespace System.Xml
 {
     public partial class XmlSecureResolver : XmlResolver
     {
-        public override Task<object> GetEntityAsync(Uri absoluteUri, string role, Type ofObjectToReturn)
+        public override Task<object> GetEntityAsync(Uri absoluteUri, string? role, Type ofObjectToReturn)
```
Comment (buyaa-n) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlSecureResolverAsync.cs:12]: Seems `Type ofObjectToReturn` can be null?

**Review by buyaa-n — APPROVED** [review]
Left few comments for public APIs, looks good thanks!

File: src/libraries/System.Private.Xml/src/System/Xml/Cache/XPathDocumentView.cs [hunk 12]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
 #if ENABLEDATABINDING
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Cache/XPathDocumentView.cs:6]: I thought about it too and conclusion was: if nothing breaks then it is most likely dead code

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEventCache.cs [hunk 13]
```diff
@@ -206,23 +207,23 @@ public void EventsToWriter(XmlWriter writer)
...
                                 writer.WriteFullEndElement();
                             break;
 
                         case XmlEventType.Nmsp:
                             if (rawWriter != null)
-                                rawWriter.WriteNamespaceDeclaration(page[idxEvent].String1, page[idxEvent].String2);
+                                rawWriter.WriteNamespaceDeclaration(page[idxEvent].String1!, page[idxEvent].String2!);
                             else
-                                writer.WriteAttributeString("xmlns", page[idxEvent].String1, XmlReservedNs.NsXmlNs, page[idxEvent].String2);
+                                writer.WriteAttributeString("xmlns", page[idxEvent].String1!, XmlReservedNs.NsXmlNs, page[idxEvent].String2!);
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEventCache.cs:226]: IIRC the ! might be there only because we use non-generic collections

[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEventCache.cs:226]: I'll check

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/BitSet.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/BitSet.cs:201]: actually just checked it and they are mutating when BitSet is growing in size

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/CompiledidEntityConstraint.cs [hunk 14]
```diff
@@ -14,8 +15,8 @@ internal class CompiledIdentityConstraint
     {
         internal XmlQualifiedName name = XmlQualifiedName.Empty;
         private readonly ConstraintRole _role;
-        private readonly Asttree _selector;
-        private readonly Asttree[] _fields;
+        private readonly Asttree _selector = null!;
+        private readonly Asttree[] _fields = null!;
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/CompiledidEntityConstraint.cs:19]: It's used in other files but it only seems to be used in case of error or circular reference - this doesn't seem to be worth changing for the error case.

I'll leave as is

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs (hunk 1 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/SchemaDeclBase.cs:36]: Ok, this logic is super convoluted but this is actually nullable

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs:14]:
```csharp
        internal void SetAttributes(XmlSchemaObjectCollection newAttributes)
        {
            _attributes = newAttributes;
        }
```

doesn't do any validation - will check if this can be non-nullable

File: src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs (hunk 6 above)
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlEncoding.cs:283]: just double checked - those are only used internally for decoding (only when decoding is auto-detected)

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEventCache.cs [hunk 15]
```diff
@@ -147,56 +148,56 @@ public void EventsToWriter(XmlWriter writer)
                             break;
 
                         case XmlEventType.Comment:
-                            writer.WriteComment(page[idxEvent].String1);
+                            writer.WriteComment(page[idxEvent].String1!);
                             break;
 
                         case XmlEventType.PI:
-                            writer.WriteProcessingInstruction(page[idxEvent].String1, page[idxEvent].String2);
+                            writer.WriteProcessingInstruction(page[idxEvent].String1!, page[idxEvent].String2!);
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEventCache.cs:155]: correct, seems no errors with this change

File: src/libraries/System.Private.Xml/src/System/Xml/Resolvers/XmlPreloadedResolverAsync.cs [hunk 16]
```diff
@@ -16,15 +17,15 @@ namespace System.Xml.Resolvers
     public partial class XmlPreloadedResolver : XmlResolver
     {
         public override Task<object> GetEntityAsync(Uri absoluteUri,
-                                             string role,
+                                             string? role,
                                              Type ofObjectToReturn)
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/Resolvers/XmlPreloadedResolverAsync.cs:21]: right, the non async version has it nullable and async also should match it (also implementation confirms)

**Review by buyaa-n — APPROVED** [review]
Updates looks good, thanks!

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlSchemaComplexContentRestriction.cs]: How about adding an assert and triggering a CI run to see if anything breaks?

**Comment by krwq** [comment]
the failures seem unrelated and infra team seems to know about them
